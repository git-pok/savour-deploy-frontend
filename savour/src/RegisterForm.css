.RegisterForm-bg-img {
  display: block;
  float: left;
  width: 100vw;
  height: 100vh;
  margin: 0px 0px;
}

.RegisterForm {
  display: flex;
  align-items: center;
  justify-content: center;
  flex-wrap: wrap;
  width: 80%;
  margin: 100px auto 0px auto;
  padding: 50px 0px;
  background-color: rgb(78, 42, 42);
  border-radius: 10px;
}

.RegisterForm-field {
  display: block;
}
  
.RegisterForm-h1 {
  text-align: center;
  display: block;
  font-size: 60px;
  margin: 0px 0px;
  padding-top: 120px;
  color:  rgb(255, 255, 255);
}

.RegisterForm-field label {
  display: block;
  width: 260px;
  height: 30px;
  margin-right: 30px;
  color: rgb(255, 255, 255);
}
  
.RegisterForm-field input {
  width: 260px;
  height: 30px;
  margin-bottom: 30px;
}

.RegisterForm-submit {
  display: block;
  width: 100%;
  margin-top: 10px;
}

.RegisterForm-submit button {
  display: block;
  width: 260px;
  height: 30px;
  margin: 0px auto;
  margin-top: 10px;
  border: 3px inset rgb(223, 216, 85);
  background-color: rgb(223, 216, 85);
  color: rgb(0, 0, 0);
  transition: .5s ease-in-out background-color;
}

.RegisterForm button:hover {
  border-color: rgb(255, 255, 255);
  background-color: rgb(255, 255, 255);
  color: rgb(0, 0, 0);
}