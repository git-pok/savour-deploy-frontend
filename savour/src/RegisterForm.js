import { useState, useEffect, useContext } from 'react';
import { useHistory } from 'react-router-dom';
import Message from './Message.js';
import SavourApi from './models/SavourApi.js';
import useLocalStorage from './hooks/useLocalStorage.js';
import useToggleState from './hooks/useToggleState.js';
import UserContext from './context/UserContext.js';
import jwt_decode from 'jwt-decode';
import image from './img/ambient-kitchen.jpg';
import './RegisterForm.css';

/**
 * RegisterForm
 * Props: formValsCheck
 *    formValsCheck: function to check if
 *    form props are missing.
 * Renders register form.
*/
const RegisterForm = ({ formValsCheck }) => {
  const initialState = {
      username: "", first_name: "", last_name: "",
      email: "", phone: "", password: ""
  };
  const { usrData, setUsrData } = useContext(UserContext);
  const [ formErrMsg, setFormErrMsg ] = useState(null);
  const [ formData, setFormData ] = useState(initialState);
  const [ formCmplt, setFormCmplt ] = useToggleState(false);
  const [ isSubmitted, setIsSubmitted ] = useToggleState(false);
  const [ invalidForm, setInvalidForm ] = useToggleState(false);
  const history = useHistory();

  useEffect(() => {
    const signup = async () => {

      try {
        const { 
          username, email, first_name,
          last_name, phone, password
        } = formData;

        const data = { 
            username, first_name, last_name, email,
            phone, password
        };
        const regResult = await SavourApi.request("post", "/users/register", data);
        const token = regResult.data[0].token;
        SavourApi.token = token;
        const payload = await jwt_decode(token);
        payload.token = token;
        payload.userUsername = payload.username;
        delete payload.username;
        payload.userId = payload.id;
        delete payload.id;
        setUsrData(data => (
          payload
        ));
        // Set isSubmitted to false.
        setIsSubmitted();
        setFormData(() => initialState);
        history.push("/");

      } catch (err) {
        console.log("ERROR", err);
        const error = err.response.data.error.message;
        const isErrObj = typeof err.message === "object";
        const isErrorObj = typeof error === "object";
        const errMsg = isErrObj ? null : err.message;
        const errorMsg = isErrorObj ? null : error;
        setFormErrMsg(() => errorMsg || errMsg || "Error!");
        setInvalidForm();
        setTimeout(setInvalidForm, 3000);
        // Set isSubmitted to false.
        setTimeout(setIsSubmitted, 3000);
      }
    }

    if (isSubmitted) signup();

  }, [isSubmitted])

  const handleChange = (evt) => {
    const { name, value } = evt.target;

    setFormData(data => ({
      ...data,
      [name]: value
    }))
  }

  const handleSubmit = async (evt) => {
    evt.preventDefault();
    const reqProps = [
      "username", "first_name", "last_name",
      "email", "phone", "password"
    ];
    // Checks if form props are missing.
    const isValMsn = formValsCheck(reqProps, formData);
    // If props are missing set invalidForm.
    if (isValMsn) {
      setFormErrMsg(() => "All fields must be complete!");
      setInvalidForm();
      setTimeout(setInvalidForm, 3000);
    }
    // Set isSubmitted to true if all props exist.
    else setIsSubmitted();
  }
  
  // image styles.
  const styles = {
    backgroundImage: `url(${image})`,
    backgroundSize: "cover",
    backgroundRepeat: "no-repeat"
  }

  return (
    <>
    <div className="RegisterForm-bg-img" style={styles}>
    <h1 className="RegisterForm-h1">Sign Up</h1>
    <form onSubmit={handleSubmit} className="RegisterForm">
      <div className="RegisterForm-field">
        <label htmlFor="username">Username</label>
        <input
            type="text"
            id="username"
            onChange={handleChange}
            value={formData.username}
            name="username"
            placeholder="Type a username"
            autoComplete="username"></input>
      </div>
      <div className="RegisterForm-field">
        <label htmlFor="firstName">First Name</label>
        <input
            type="text"
            id="firstName"
            onChange={handleChange}
            value={formData.first_name}
            name="first_name"
            placeholder="Type a first name"></input>
      </div>
      <div className="RegisterForm-field">
        <label htmlFor="lastName">Last Name</label>
        <input
            type="text"
            id="lastName"
            onChange={handleChange}
            value={formData.last_name}
            name="last_name"
            placeholder="Type a last name"></input>
      </div>
      <div className="RegisterForm-field">
        <label htmlFor="email">Email</label>
        <input
            type="email"
            id="email"
            onChange={handleChange}
            value={formData.email}
            name="email"
            placeholder="Type an email"></input>
      </div>
      <div className="RegisterForm-field">
        <label htmlFor="phone">Phone</label>
        <input
            type="text"
            id="phone"
            onChange={handleChange}
            value={formData.phone}
            name="phone"
            placeholder="Type a phone"></input>
      </div>
      <div className="RegisterForm-field">
        <label htmlFor="password">Password</label>
        <input
            type="password"
            id="password"
            onChange={handleChange}
            value={formData.password}
            name="password"
            placeholder="Type a password"
            autoComplete="current-password"></input>
      </div>
      <div className="RegisterForm-submit">
      {
        invalidForm &&
        <Message msgObj={
          {
            class: "fail",
            msg: formErrMsg ? formErrMsg : "Error!"
          }
        } />
      }
        <button>SUBMIT</button>
      </div>
    </form>
    </div>
    </>
  );
}

export default RegisterForm;