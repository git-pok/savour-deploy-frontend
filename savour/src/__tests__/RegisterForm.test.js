import RegisterForm from '../RegisterForm.js';
import { render } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';
import UserContext from '../context/UserContext.js';


const usrData = {
  userUsername: "fvin",
  userId: 12
};


test("renders RegisterForm without crashing", () => {
  render(
    <MemoryRouter>
      <UserContext.Provider value={{usrData}}>
        <RegisterForm />
      </UserContext.Provider>
    </MemoryRouter>
  );
});

test("renders snapshot without crashing", () => {
  const { asFragment } = render(
    <MemoryRouter>
    <UserContext.Provider value={{usrData}}>
      <RegisterForm />
    </UserContext.Provider>
    </MemoryRouter>
  );

  expect(asFragment()).toMatchSnapshot();
});