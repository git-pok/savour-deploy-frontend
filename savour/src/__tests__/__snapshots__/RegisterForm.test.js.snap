// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders snapshot without crashing 1`] = `
<DocumentFragment>
  <div
    class="RegisterForm-bg-img"
    style="background-image: url(ambient-kitchen.jpg); background-size: cover; background-repeat: no-repeat;"
  >
    <h1
      class="RegisterForm-h1"
    >
      Sign Up
    </h1>
    <form
      class="RegisterForm"
    >
      <div
        class="RegisterForm-field"
      >
        <label
          for="username"
        >
          Username
        </label>
        <input
          autocomplete="username"
          id="username"
          name="username"
          placeholder="Type a username"
          type="text"
          value=""
        />
      </div>
      <div
        class="RegisterForm-field"
      >
        <label
          for="firstName"
        >
          First Name
        </label>
        <input
          id="firstName"
          name="first_name"
          placeholder="Type a first name"
          type="text"
          value=""
        />
      </div>
      <div
        class="RegisterForm-field"
      >
        <label
          for="lastName"
        >
          Last Name
        </label>
        <input
          id="lastName"
          name="last_name"
          placeholder="Type a last name"
          type="text"
          value=""
        />
      </div>
      <div
        class="RegisterForm-field"
      >
        <label
          for="email"
        >
          Email
        </label>
        <input
          id="email"
          name="email"
          placeholder="Type an email"
          type="email"
          value=""
        />
      </div>
      <div
        class="RegisterForm-field"
      >
        <label
          for="phone"
        >
          Phone
        </label>
        <input
          id="phone"
          name="phone"
          placeholder="Type a phone"
          type="text"
          value=""
        />
      </div>
      <div
        class="RegisterForm-field"
      >
        <label
          for="password"
        >
          Password
        </label>
        <input
          autocomplete="current-password"
          id="password"
          name="password"
          placeholder="Type a password"
          type="password"
          value=""
        />
      </div>
      <div
        class="RegisterForm-submit"
      >
        <button>
          SUBMIT
        </button>
      </div>
    </form>
  </div>
</DocumentFragment>
`;
